lib/std/ring.zig

```diff
@@ -0,0 +1,304 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const Atomic = std.atomic.Atomic;
+const expectEqual = std.testing.expectEqual;
+
+pub fn Ring(comptime T: type, comptime size: usize) type {
+    return if (builtin.single_threaded)
+        RingGeneric(T, size, .non_atomic)
+    else
+        RingGeneric(T, size, .mpmc);
+}
+
+const RingKind = enum {
+    mpmc,
+    non_atomic,
+};
+
+/// A circular buffer with a tail and a head. The tail is an index within the
+/// buffer from which the next item will be read. The head is an index within
+/// the buffer to which the next item will be written. Both the tail and
+/// the head are atomically incremented via a CAS loop and are bound to the
+/// range of `0..std.math.maxInt(usize)`. The size is comptime asserted to be
+/// a power of two. This permits the tail and head to overflow while remaining
+/// aligned to the size. Every slot within the buffer is tagged with an atomic
+/// enum indicating the current state of the slot (occupied or empty).
+///
+/// *Credit*: This draws upon the Linux Kernel's BPF ring buffer specification
+/// (but is a novel implementation). Specifically, the idea of tagged slots and
+/// of monotonic, wrapping head/tail.
```

Comment:
wrapping head/tail is a standard approach as it compiles down to unconditionally applying a bitmask on the index (you could also compute and write it down explicitly).

More of a curiosity: How does the approach compare to [this paper](https://www.osti.gov/servlets/purl/1531271) ?
As I understand it, this approach is optimized to use less memory and less complex than the Kernel approach of an additional data structure to track the status.

Reply:
> wrapping head/tail with 2s complement is a standard approach

Blew my mind when I read it. It probably is standard. crossbeam and DPDK appear to use it.

> How does the approach compare to [this paper](https://www.osti.gov/servlets/purl/1531271) ?

I hadn't seen that paper until now. I think you're incorrect in saying that paper doesn't utilize "an additional data structure to track the status." From an initial look, it seems like they propose stamping each node within the buffer with a "seqid" (end of Section 4 Paragraph 2). It actually seems to be the inspiration or direct attribution of Rust's [crossbeam::ArrayQueue](https://docs.rs/crossbeam/latest/crossbeam/queue/struct.ArrayQueue.html). There they use the upper bits of the head/tail as the "lap."

I'll have to sit with this and digest it for a bit. Do you think re-designing around a different algorithm now would be preferable to merging the existing design?

Reply:
@matu3ba, your comment made me look at some other implementations. I really liked DPDK's method since it permitted untagged slots. I scrapped everything and re-implemented using their producer/consumer spans method. I still haven't given that paper a thorough read...